### PR TITLE
ARROW-11247: [C++] Infer date32 columns in CSV

### PR DESCRIPTION
**Pull request by nealrichardson**

### CODE REVIEW

File: cpp/src/arrow/csv/column_builder_test.cc [hunk 1]
```diff
@@ -15,13 +15,14 @@
 // specific language governing permissions and limitations
 // under the License.
 
+#include "arrow/csv/column_builder.h"
+
+#include <gtest/gtest.h>
+
```
Comment (nealrichardson) [inline cpp/src/arrow/csv/column_builder_test.cc:21]: My editor runs clang-format and IDK why it moved this

**Comment by nealrichardson** [comment]
Some python tests are failing with this change; I think the tests should be updated from where it assumes dates will be parsed (suboptimally IMO) as timestamps, but @jorisvandenbossche maybe you can review and weigh in.

**Comment by jorisvandenbossche** [comment]
I added a commit that updates the tests for the new behaviour, *in case* we decide we are OK with that.

Generally I think we should do the best inference from Arrow's point of view, and which is thus date type for a date string. 

The only reason I am thinking to not do it is that, for people converting the data to pandas afterwards, dates are not that well supported (at this point) in pandas. Now, there is a `to_pandas(..., date_as_object=False)` keyword a user can specify to still get a datetime64 dtype in pandas instead of datetime.date objects.

**Comment by pitrou** [comment]
Please update the docs for https://arrow.apache.org/docs/cpp/csv.html#data-types and https://arrow.apache.org/docs/python/csv.html

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Travis-CI build: https://travis-ci.com/github/pitrou/arrow/builds/213173530

**Pull request closed without merging**